File: tools/generate-sitemap.mjs
import fs from "fs";
import path from "path";

const BASE_URL = "https://www.sindikatstudio83.me";
const ROOT_DIR = path.resolve("Pages");
const OUT_FILE = path.join(ROOT_DIR, "sitemap.xml");

// Fajlovi koje preskačemo
const SKIP_FILES = new Set(["404.html"]);

// Folderi koje preskačemo
const SKIP_DIRS = new Set([
  ".git",
  ".github",
  "node_modules",
  "tools",
  "_templates",
]);

function isSkippableDir(dirName) {
  return SKIP_DIRS.has(dirName);
}

function normalizePathSlashes(p) {
  return ("/" + p).replace(/\/{2,}/g, "/");
}

// ✅ Samo clean URL-ovi: index.html -> folder URL
// ❌ Bilo koji drugi .html (legacy ili stub) -> null (ne ulazi u sitemap)
function fileToUrl(filePath) {
  let rel = path.relative(ROOT_DIR, filePath).replace(/\\/g, "/");
  const lower = rel.toLowerCase();

  // ⛔️ Safety: ako se .html provuče kao dio foldera (npr. /usluge.html/index.html), izbaci
  if (lower.includes(".html/")) return null;

  // DOZVOLI SAMO index.html
  const isRootIndex = lower === "index.html";
  const isFolderIndex = lower.endsWith("/index.html");
  if (!isRootIndex && !isFolderIndex) return null;

  if (isRootIndex) rel = ""; // root
  else rel = rel.slice(0, -"/index.html".length) + "/";

  let finalPath = normalizePathSlashes(rel);

  // osiguraj trailing slash na folder rutama (osim root-a koji je "/")
  if (finalPath !== "/" && !finalPath.endsWith("/")) finalPath += "/";

  return BASE_URL + finalPath;
}

// Pravila za izbacivanje URL-ova koje ne želimo u sitemap (safety net)
function shouldSkipUrl(url) {
  if (!url) return true;

  const lower = url.toLowerCase();

  // 1) izbaci blog template placeholder
  if (lower.includes("/sr-me/blog/<slug>/")) return true;
  if (lower.includes("/sr-me/blog/%3cslug%3e/")) return true;

  // 2) izbaci internal/template putanje
  if (lower.includes("/_templates/")) return true;

  // 3) izbaci test rute (ako ih ima)
  if (lower.includes("/test/")) return true;

  // 4) zaštita: ako se ikad provuče ".html" BILO GDJE u putanji, izbaci ga
  // hvata i ".../neko.html/" i ".../neko.html"
  if (/\.html(\/|$)/i.test(url)) return true;

  return false;
}

function walk(dir) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  let files = [];

  for (const e of entries) {
    const full = path.join(dir, e.name);

    if (e.isDirectory()) {
      if (!isSkippableDir(e.name)) {
        files = files.concat(walk(full));
      }
      continue;
    }

    if (!e.isFile()) continue;
    if (!e.name.toLowerCase().endsWith(".html")) continue;
    if (SKIP_FILES.has(e.name)) continue;

    files.push(full);
  }

  return files;
}

function isoDateFromMtime(filePath) {
  const st = fs.statSync(filePath);
  return st.mtime.toISOString().slice(0, 10); // YYYY-MM-DD
}

function escapeXml(s) {
  return String(s || "")
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&apos;");
}

function buildSitemap(urlEntries) {
  const header =
    `<?xml version="1.0" encoding="UTF-8"?>\n` +
    `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n`;

  const body = urlEntries
    .map(({ loc, lastmod }) => {
      return (
        `  <url>\n` +
        `    <loc>${escapeXml(loc)}</loc>\n` +
        (lastmod ? `    <lastmod>${lastmod}</lastmod>\n` : "") +
        `  </url>\n`
      );
    })
    .join("");

  return header + body + `</urlset>\n`;
}

function main() {
  if (!fs.existsSync(ROOT_DIR)) {
    console.error(`ERROR: Pages folder not found at: ${ROOT_DIR}`);
    process.exit(1);
  }

  const htmlFiles = walk(ROOT_DIR);

  // napravi entries samo iz index.html fajlova
  const entries = htmlFiles
    .map((fp) => {
      const loc = fileToUrl(fp);
      if (!loc) return null;
      return {
        loc,
        lastmod: isoDateFromMtime(fp),
        file: fp,
      };
    })
    .filter(Boolean)
    .filter((x) => !shouldSkipUrl(x.loc));

  // ukloni duplikate
  const seen = new Set();
  const unique = [];
  for (const e of entries) {
    if (seen.has(e.loc)) continue;
    seen.add(e.loc);
    unique.push(e);
  }

  unique.sort((a, b) => a.loc.localeCompare(b.loc));

  const xml = buildSitemap(unique);
  fs.writeFileSync(OUT_FILE, xml, "utf8");

  console.log(`✅ sitemap generated: ${OUT_FILE}`);
  console.log(`✅ URLs: ${unique.length}`);
}

main();

File: tools/generate_sitemap.py
#!/usr/bin/env python3
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
import re
from xml.dom import minidom
from xml.etree.ElementTree import Element, SubElement, tostring

ROOT = Path(__file__).resolve().parents[1]
PAGES = ROOT / "Pages"
BASE_URL = "https://www.sindikatstudio83.me"

EXCLUDE_PARTS = {"<slug>", "_templates", "assets"}


def to_url(path: Path) -> str:
    rel = path.relative_to(PAGES).as_posix()
    if rel == "index.html":
        return BASE_URL + "/"
    if rel.endswith("/index.html"):
        rel = rel[:-10]
    rel = "/" + rel.lstrip("/")
    rel = re.sub(r"/+", "/", rel)
    if not rel.endswith(".html") and not rel.endswith("/"):
        rel += "/"
    return f"{BASE_URL}{rel}"


def priority_for(url: str) -> str:
    explicit = {
        f"{BASE_URL}/": "1.0",
        f"{BASE_URL}/sr-me/": "0.9",
        f"{BASE_URL}/sr-me/pocetna/": "0.9",
        f"{BASE_URL}/sr-me/usluge/": "0.9",
        f"{BASE_URL}/sr-me/blog/": "0.9",
        f"{BASE_URL}/sr-me/onama/": "0.8",
        f"{BASE_URL}/sr-me/kontakt/": "0.8",
        f"{BASE_URL}/sr-me/poslovi/": "0.8",
    }
    if url in explicit:
        return explicit[url]
    if "/blog/p" in url:
        return "0.7"
    if "/usluge/" in url:
        return "0.8"
    if "/onama/cn.html" in url:
        return "0.5"
    return "0.8"


def changefreq_for(url: str) -> str:
    if "/blog/p" in url:
        return "monthly"
    if "/blog/" in url:
        return "daily"
    if "/onama/cn.html" in url:
        return "yearly"
    if url in {f"{BASE_URL}/", f"{BASE_URL}/sr-me/", f"{BASE_URL}/sr-me/pocetna/", f"{BASE_URL}/sr-me/usluge/", f"{BASE_URL}/sr-me/poslovi/"}:
        return "weekly"
    return "monthly"


def should_include(file: Path) -> bool:
    parts = set(file.parts)
    if EXCLUDE_PARTS & parts:
        return False
    text = file.read_text(encoding="utf-8", errors="ignore").lower()
    if 'name="robots" content="noindex' in text:
        return False
    return True


def iso_lastmod(file: Path) -> str:
    ts = file.stat().st_mtime
    return datetime.fromtimestamp(ts, tz=timezone.utc).date().isoformat()


def main() -> None:
    html_files = sorted(p for p in PAGES.rglob("*.html") if should_include(p))
    urlset = Element("urlset", xmlns="http://www.sitemaps.org/schemas/sitemap/0.9")

    for file in html_files:
        url = to_url(file)
        u = SubElement(urlset, "url")
        SubElement(u, "loc").text = url
        SubElement(u, "lastmod").text = iso_lastmod(file)
        SubElement(u, "changefreq").text = changefreq_for(url)
        SubElement(u, "priority").text = priority_for(url)

    rough = tostring(urlset, encoding="utf-8")
    pretty = minidom.parseString(rough).toprettyxml(indent="  ", encoding="utf-8").decode("utf-8")
    pretty = "\n".join(line for line in pretty.splitlines() if line.strip()) + "\n"

    out = PAGES / "sitemap.xml"
    out.write_text(pretty, encoding="utf-8")
    print(f"Generated {out.relative_to(ROOT)} with {len(html_files)} URLs")


if __name__ == "__main__":
    main()
